### DISCUSSION TITLE
Git LFS fails if git-lfs-authenticate bearer token expires

### DISCUSSION THREAD

**Discussion by Uxorious** [git-lfs/git-lfs#5814] [General]
I'm seeing a problem cloning repositories with a lot of LFS files using SSH over slow connections.
Basically the https token given by git-lfs-authenticate will expire before all the files are downloaded, and then it will stop.
Is there any way to get it to re-request a token if it expires?

**Comment by chrisd8088** [comment]
Hey, I'm sorry you're having trouble.  Would you be able to provide more detail, particularly the output of `git lfs env` (when run from inside the local repository clone) and a trace log of your Git clone operation when run with the `GIT_TRACE=1 GIT_TRANSFER_TRACE=1 GIT_CURL_VERBOSE=1` environment variables set?

You also don't mention which Git LFS hosting service you're using, but if it's a commercial one, you may want to contact their support team for assistance.

I wasn't able to replicate your problem when using GitHub as a Git LFS hosting service and injecting an explicit timeout into the Git LFS Batch API [request](https://github.com/git-lfs/git-lfs/blob/fc61febe9cc2d9ddc6ffe3e8d1ae546512632552/tq/api.go#L86-L98) so that the first batch API request succeeds, and then the Git LFS client would wait 11 minutes before making its second batch API request, at which point the token provided by `git-lfs-authenticate` has expired (because GitHub assigns a 10-minute expiry time to their tokens).

When this occurs, the second batch API request receives a 403 from the GitHub server, which prompts the client to run `git-lfs-authenticate` again, and then retry the batch API request with the refreshed token.  After that, the clone proceeds normally and all Git LFS objects are downloaded.

```
10:46:46.688823 trace git-lfs: run_command: ssh git@github.com git-lfs-authenticate chrisd8088/test.git download
10:46:46.689032 trace git-lfs: exec: ssh 'git@github.com' 'git-lfs-authenticate chrisd8088/test.git download'
...
10:46:48.017576 trace git-lfs: HTTP: POST https://lfs.github.com/chrisd8088/test/objects/batch
> POST /chrisd8088/test/objects/batch HTTP/1.1
> Host: lfs.github.com
...
< HTTP/2.0 200 OK
...
10:57:48.646711 trace git-lfs: HTTP: POST https://lfs.github.com/chrisd8088/test/objects/batch
> POST /chrisd8088/test/objects/batch HTTP/1.1
> Host: lfs.github.com
...
< HTTP/2.0 403 Forbidden
...
10:57:48.976938 trace git-lfs: ssh cache expired: git@github.com git-lfs-authenticate chrisd8088/test.git download
10:57:48.976986 trace git-lfs: run_command: ssh git@github.com git-lfs-authenticate chrisd8088/test.git download
10:57:48.977340 trace git-lfs: exec: ssh 'git@github.com' 'git-lfs-authenticate chrisd8088/test.git download'
...
10:57:50.196661 trace git-lfs: HTTP: POST https://lfs.github.com/chrisd8088/test/objects/batch
> POST /chrisd8088/test/objects/batch HTTP/1.1
> Host: lfs.github.com
...
< HTTP/2.0 200 OK
```

From that trace you can see the client's cache of the credentials provided by `git-lfs-authenticate` is [reported](https://github.com/git-lfs/git-lfs/blob/fc61febe9cc2d9ddc6ffe3e8d1ae546512632552/lfshttp/ssh.go#L44-L45) as expired, and the command is run again to retrieve a fresh token.

So I'm not sure what's going on for you, but if you generate and review your own trace logs (and provide them here if you're still stuck) that may help clarify the situation.

**Comment by Uxorious** [comment]
Thanks for the reply.

This is using Artifactory as the LFS provider (repo itself is on GHE).

Env is really simple, so not really worth posting here.

I guess I'll try going to Artifactory for more info.

But in case it's useful, I am never seeing the "ssh cache expired" trace messages.
After it's been running for a while, it makes an HTTP POST to get the list of the next 100 files.
This succeeds with a 200 and returns a big list of URLs of the 100 files, ending in:
```
12:43:17.873921 trace git-lfs: xfer: adapter "basic" End()
12:43:17.873921 trace git-lfs: xfer: adapter "basic" worker 0 stopping
12:43:17.873921 trace git-lfs: xfer: adapter "basic" stopped
```
Then it starts working on the actual downloads:
```
12:43:17.873921 trace git-lfs: tq: starting transfer adapter "basic"
12:43:17.873921 trace git-lfs: xfer: adapter "basic" Begin() with 1 workers
12:43:17.873921 trace git-lfs: xfer: adapter "basic" started
12:43:17.873921 trace git-lfs: xfer: adapter "basic" worker 0 starting
12:43:17.874444 trace git-lfs: xfer: adapter "basic" worker 0 processing job for "dbac78dfda54965f6a13431070d63024b9112c914d16ecc1d13ab8172fceeaae"
12:43:17.874972 trace git-lfs: HTTP: GET https://artifactory...
```

It successfully gets a handful of files, and then gets a 401.
But it just continues trying to get files from that list of 100, resulting in mostly 401's but oddly also a 200 here and there.
Then it just goes crazy with POST commands (all failing with 401s) to go through the rest of the 1000's of files.

So, it really looks like it never tries to renew the token.

**Comment by chrisd8088** [comment]
> It successfully gets a handful of files, and then gets a 401.
> But it just continues trying to get files from that list of 100, resulting in mostly 401's but oddly also a 200 here and there.

One thing to note, BTW, is that the token from the `git-lfs-authenticate` command is not used for individual object file downloads.  The JSON response from the Batch API request (the `POST` request) will [contain](https://github.com/git-lfs/git-lfs/blob/fc61febe9cc2d9ddc6ffe3e8d1ae546512632552/docs/api/batch.md#successful-responses) `header` fields for each individual object, and those will likely be signed and will each have some kind of expiry time assigned.

Do you see `git-lfs-authenticate` at all in your trace logs?  (I ask because sometimes one has an SSH Git remote, but the repo contains an `.lfsconfig` file which defines an HTTPS remote URL for Git LFS, and then the Git LFS client doesn't run `git-lfs-authenticate` at all.)

**Reply by Uxorious** [reply]
> Do you see git-lfs-authenticate at all in your trace logs?

Yes, I see some WAY early in the clone - like 30 seconds after it started.

At first, right after it starts the filter-process, I get:
```
17:16:34.782357 trace git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'remote'
17:16:34.814615 trace git-lfs: tq: running as batched queue, batch size of 100
17:16:34.816167 trace git-lfs: filepathfilter: accepting "blah/filename.bin"
17:16:34.816690 trace git-lfs: attempting pure SSH protocol connection
17:16:34.816690 trace git-lfs: spawning pure SSH connection
17:16:34.816690 trace git-lfs: run_command: ssh git@artifactory.company.com git-lfs-transfer /artifactory/gitlfs-local download
17:16:34.822959 trace git-lfs: exec: ssh 'git@artifactory.company.com' 'git-lfs-transfer /artifactory/gitlfs-local download'
17:16:35.356149 trace git-lfs: pure SSH connection successful
17:16:35.356149 trace git-lfs: pure SSH protocol connection failed: Unable to negotiate version with remote side (unable to read capabilities): EOF
```

Then there's a bunch more filepathfilter stuff, and then:
```
17:16:35.499766 trace git-lfs: tq: sending batch of size 100
17:16:35.500272 trace git-lfs: filepathfilter: accepting "blah/otherfilename.bin"
17:16:35.500272 trace git-lfs: run_command: ssh git@artifactory.company.com git-lfs-authenticate /artifactory/gitlfs-local download
[SNIP more filepathfilter stuff]
17:16:35.509723 trace git-lfs: exec: ssh 'git@artifactory.company.com' 'git-lfs-authenticate /artifactory/gitlfs-local download'
```

> (I ask because sometimes one has an SSH Git remote, but the repo contains an .lfsconfig file which defines an
> HTTPS remote URL for Git LFS, and then the Git LFS client doesn't run git-lfs-authenticate at all.)

Yes this repo does indeed have an .lfsconfig file (since LFS isn't hosted on GH, it makes possible to clone without having to remember the LFS URL).
And it is indeed HTTPS, but I'm using a URL rewrite rule to fix that:
```
git config --global url.ssh://git@artifactory.company.com/artifactory/.insteadOf https://artifactory.company.com/artifactory/api/lfs/
```

Could that be causing issues?

**Reply by chrisd8088** [reply]
> And it is indeed HTTPS, but I'm using a URL rewrite rule to fix that:
> Could that be causing issues?

I wouldn't think so, since you're seeing an initially successful authentication and subsequent requests.

I would suggest focusing on the Batch API requests, i.e., the `POST` requests, to see why they aren't retried after the token expires.

One thing to check might be to check what your Artifactory server is returning as an expiry time, and if that corresponds to the point where it starts returning 401s.  If you manually run:
```
$ ssh 'git@artifactory.company.com' 'git-lfs-authenticate /artifactory/gitlfs-local download
```
that should give you a response with an `expires_at` and/or `expires_in` JSON field, with expiry periods in seconds.  Does that time interval match the point in your trace log, following the initial successful requests, where you start to see 401 responses to the `POST` requests to the Batch API?  (To put it another way, you want to rule out whether Artifactory is returning 401s before the client thinks the token has expired.  If that happened, it would explain why the client doesn't attempt to refresh the token.)

Another test, if it's possible, might be to have Artifactory return 403s instead of 401s, just in case there's something in the Git LFS client which is responding differently between those two cases.  I don't see anything obvious which would result the behaviour you describe, but it could still be worth testing.

Finally, what do you see in your trace logs after a `POST` request receives a 401?  Here's what I see, after a 403:

```
< HTTP/2.0 403 Forbidden
...
10:57:48.975080 trace git-lfs: HTTP: {"message":"Bad credentials","documentation_url":"https://support.github.com/contact","status":"403"}
{"message":"Bad credentials","documentation_url":"https://support.github.com/contact","status":"403"}
10:57:48.975247 trace git-lfs: api error: Bad credentials
10:57:48.975357 trace git-lfs: xfer: adapter "basic" End()
10:57:48.975454 trace git-lfs: xfer: adapter "basic" worker 1 stopping
10:57:48.975472 trace git-lfs: xfer: adapter "basic" worker 4 stopping
10:57:48.975470 trace git-lfs: xfer: adapter "basic" worker 3 stopping
10:57:48.975488 trace git-lfs: xfer: adapter "basic" worker 5 stopping
10:57:48.975497 trace git-lfs: xfer: adapter "basic" worker 6 stopping
10:57:48.975502 trace git-lfs: xfer: adapter "basic" worker 0 stopping
10:57:48.975482 trace git-lfs: xfer: adapter "basic" worker 7 stopping
10:57:48.975493 trace git-lfs: xfer: adapter "basic" worker 2 stopping
10:57:48.975533 trace git-lfs: xfer: adapter "basic" stopped
10:57:48.976168 trace git-lfs: tq: running as batched queue, batch size of 100
10:57:48.976464 trace git-lfs: filepathfilter: accepting "foo.bin"
Downloading foo.bin (40 B)
10:57:48.976611 trace git-lfs: tq: running as batched queue, batch size of 100
10:57:48.976663 trace git-lfs: tq: sending batch of size 100
10:57:48.976938 trace git-lfs: ssh cache expired: git@github.com git-lfs-authenticate chrisd8088/test.git download
10:57:48.976986 trace git-lfs: run_command: ssh git@github.com git-lfs-authenticate chrisd8088/test.git download
10:57:48.977340 trace git-lfs: exec: ssh 'git@github.com' 'git-lfs-authenticate chrisd8088/test.git download'
10:57:50.196268 trace git-lfs: api: batch 1 files
10:57:50.196661 trace git-lfs: HTTP: POST https://lfs.github.com/chrisd8088/test/objects/batch
> POST /chrisd8088/test/objects/batch HTTP/1.1
```

**Reply by Uxorious** [reply]
> One thing to check might be to check what your Artifactory server is returning as an expiry time, and if that corresponds to the point where it starts returning 401s. If you manually run:
> `$ ssh 'git@artifactory.company.com' 'git-lfs-authenticate /artifactory/gitlfs-local download`

Interesting - that is significantly different from GitHub.
GitHub gives me a JSON like you said, but Artifactory gives me something like:
`{"header":{"Authorization":"Bearer LOTS_OF_BASE64"},"href":"https://artifactory.company.com/artifactory/api/lfs/gitlfs-local"}
`
And if I decode it, I get a JWT.

I'll see if I can get somewhere with Artifactory support.
Thanks a ton so far!

**Reply by chrisd8088** [reply]
Interesting.  As you may have already checked, GitHub, for instance, provides something like the following to a `git-lfs-authenticate` request:
```json
{
  "href": "https://lfs.github.com/...",
  "header": {
    "Authorization": "RemoteAuth gitauth-v1-..."
  },
  "expires_at": "2024-07-03T05:55:50Z",
  "expires_in": 599
}
```

If Artifactory isn't offering those `expires_*` fields, then the Git LFS client won't know when the token expires, and so likely won't ask to refresh it, but will take 4xx responses to mean the client is just unauthorized and should not try again.  (After all, we don't want the client to retry endlessly in response to a 4xx code.)
